Shade Groups When All Sub-Rows Complete In Refl GUI

cailafinn opened the pull request:
**Description of work.**
This PR adds a pointer from each row to the parent group to allow the group to update its shade when all its children are complete. That's it. 

So why is it so big?

  Turns out the groups get moved and copied around a lot. This meant that the pointer would often point to the wrong memory address if the group was moved or copied. This had a tendancy to cause weird bugs/segfaults, so new copy and move constructors were created to ensure that groups updated all their children whenever they were moved. 
  The process also included a lot of trial and error. In attempting to remove the copy constructor, a lot of maintenance was done removing unnecessary copies where they occured. We had to put the copy constructor back in the end. But the work done removing the copies was still worth keeping.  

**To test:**
1. Ensure you have the ISIS sample data downloaded
1. Open the ISIS Reflectometry GUI
2. Load a Batch (top left) from the sample data, and click the `+` button to expand all groups
3. Click each run within the first group and click process *one at a time*
4. You should see the highlight colour of the group change when both individual runs are complete
5. On the next group, select one run and process it. 
6. Select the other run and delete that row
7. You should see the group change to a green highlight as all runs within it are now complete.

Fixes #26531 

<!-- Ensure the base of this PR is correct (e.g. release-next or master)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

Commit:
Add pointer to parent group for each row

It would be beneficial for a number of different issues to allow rows to
reference the parent group.

RE #26531

Commit:
Set parent when adding new rows

RE #26531

Commit:
Set all parents when loading batch

RE #26531

Commit:
Highlight groups when child rows are complete

RE #26531

Commit:
Fix broken test

The test resulted in a sort-of parent being set but was still null
underneath. This change sets the row's parent to be null explicitly so
that the checks in the changes can be made properly.

RE #26531

Commit:
Add tests for group state

RE #26531

Commit:
Remove old getParent method

This old method is an awkward workaround. It can just use the pointer
now.

RE #26531

Commit:
Attempt to remove copy Constructor

Allows us to see and potentially remove uneccessary copies of groups,
also may help fix a bug with removal of the old getParent method.

RE #26531

Commit:
Refactor to remove unneccessary copies

Deleted group copy contructor and removed unnecessary copies.
Removed unused function
Simplify clipboard tests

RE #26531

Co-authored-by: Gemma Guest <gemmaguest@users.noreply.github.com>

Commit:
Re-add copy constructor

MergeAdditionalJobs specifically requires a copy to function properly.
Could remove but would be complex as the function affects the view.

RE #26531

Co-authored-by: Gemma Guest <gemmaguest@users.noreply.github.com>

Commit:
Add row tests and mock group

Minor modifications to group header layout to separate overrides.

RE #26531

Co-authored-by: Gemma Guest <gemmaguest@users.noreply.github.com>

Commit:
Fix invalid memory address on row init

Explicitly set row's parents to be nullptr until they are set by the
group or other function.

RE #26531

Commit:
Create tests for new group functionality

Also adds tests for the new move and copy constructors I haven't written
yet.

RE #26531

Commit:
Implement custom move and copy constuctors

Due to the nature of the pointer from row to group, if the group was
moved the memory address that row held on to would be invalid. These
changes create custom move and copy constructors such that whenever the
group is moved or copied, the group gives each of its rows its new
memory address.

RE #26531

Commit:
Add release notes

RE #26531

Commit:
Clang-format

RE #26531

Commit:
Fix CppCheck constParameter failures

RE #26531

gemmaguest commented:
Moved to 6.4 because of the maintenance work involved

gemmaguest reviewed (changes requested):
There's a misunderstanding in the behaviour here where "complete" on a group should mean that post-processing is complete, but now means that child rows are complete - and therefore it will actually no longer progress to postprocess the group! An additional state is required for child rows being complete. Probably easier to chat about this in person. You've done all the difficult groundwork though so this should be easy to add now.

Otherwise, just a few minor comments but this looks really good and should make adding new features easier going forward. Good work too on removing all the unnecessary copies :+1:

Commit:
Move release notes and improve naming

RE #26531

gemmaguest commented on qt/scientific_interfaces/ISISReflectometry/Reduction/Group.cpp:
```diff
@@ -14,13 +14,32 @@
+  setAllRowParentsImpl();
+}
+
+Group::Group(Group &&old_group) noexcept : Group(old_group.m_name) {
```
Ah ok I see, it's because the default constructor requires the name

gemmaguest commented on qt/scientific_interfaces/ISISReflectometry/Reduction/Group.h:
```diff
@@ -16,53 +17,78 @@ namespace MantidQt {
+  Group &operator=(Group);
+
+  friend void swap(Group &first, Group &second) {
+    std::swap(first.m_name, second.m_name);
```
Ah I see it's the copy and swap idiom

Commit:
Improve constructor clarity and remove uneeded method

Made Group final, therefore removing the need for the Impl method.

RE #26531

Co-authored-by: Gemma Guest <gemmaguest@users.noreply.github.com>

gemmaguest commented on qt/scientific_interfaces/ISISReflectometry/Reduction/Group.cpp:
```diff
@@ -182,26 +213,48 @@ int Group::totalItems() const {
+      std::all_of(m_rows.cbegin(), m_rows.cend(), [](auto const &row) { return row && row->success(); })) {
+    setSuccess();
+  } else {
+    resetState(false);
```
Ah yes I see the state needs to be reset if we have added rows

Commit:
Add new state for all child rows being successful

RE #26531

Commit:
Refactor duplicated all_of

RE #26531

Commit:
Refactor parent knowledege exclusively into group or row

RE #26531

Commit:
Refactor parent styling update

RE #26531

Commit:
Simplidy accumulate to improve formatting

RE #26531

Commit:
Rename updateParent in group to be more accurate

RE #26531

Commit:
Fix post-processing not working for child finished state

RE #26531

Commit:
Update tests to new functionality

Progress bar tests needed the order swapping about as the status of the
group is updated after any child row is complete. This means that for
testing updates to the group's status should be done *after* any child
rows.

RE #26531

Commit:
Rename ITEM_COMPLETE

Was a bit ambiguous as complete does not necessarily mean that it was
successful. This fits more in line with the rest of the code where
complete simply means it has finished processing.

RE #26531

Commit:
Remove unneeded nullptr

RE #26531

Commit:
Update refl docs with new functionality

RE #26531

gemmaguest commented on qt/scientific_interfaces/ISISReflectometry/test/RunsTable/RunsTablePresenterProcessingTest.h:
```diff
@@ -175,6 +175,14 @@ class RunsTablePresenterProcessingTest : public CxxTest::TestSuite, RunsTablePre
+  void testRowParentStateChangedForAllRowsInGroupComplete() {
+    auto presenter = makePresenter(m_view, oneGroupWithARowModel());
+    expectRowState(Colour::SUCCESS);
+    expectGroupState(Colour::SUCCESS);
```
I'm a bit confused why this is passing - shouldn't the expected group colour be `CHILDREN_SUCCESS`? And actually we haven't even called `setSuccess` on the row so this should fail anyway? I'm having trouble running my tests locally so can't check but are these tests definitely being run?

ConorMFinn replied:
They are being run, but for some reason the EXPECT_CALL isn't working like I'd expect

gemmaguest replied:
This is an existing problem. I've raised issue #33474 to cover it because it's out of scope of this PR.

gemmaguest reviewed (changes requested):
Thanks for the changes. Looking really good now. Just a few more minor tidying points. There is actually one minor bug I've spotted with some behaviour around handling of invalid rows which I hadn't considered before. I'll make a note of that on the issue but I think it's such a minor edge case that we shouldn't hold up this PR for it.

Commit:
Make applyStylingToParent only take row objects

RE #26531

Commit:
Make minor structural fixes

RE #26531

Commit:
Fix parent state test

RE #26531

gemmaguest reviewed (approved):
Thanks for the final tidying. Code looks great now and functions as expected :+1:

DavidFair commented on qt/scientific_interfaces/ISISReflectometry/Reduction/Row.h:
```diff
@@ -48,10 +50,19 @@ class MANTIDQT_ISISREFLECTOMETRY_DLL Row : public Item {
   bool hasOutputWorkspace(std::string const &wsName) const;
   void renameOutputWorkspace(std::string const &oldName, std::string const &newName) override;
 
+  void setParent(IGroup *parent) const;
```
This shouldn't be marked const, and we should drop the mutable on `m_parent`

As a rule of thumb observable changes shouldn't have const on them, if this was an internal detail we can't see (e.g. a cache) then `const` and mutable would be fine

Otherwise people can run into undefined behaviour or bugs when they grab a `const IGroup` that changes underneath them

gemmaguest replied:
We had a lot of issues making this non-const (the whole GUI is essentially built around Rows being const!). The parent pointer is the only part of a row that can change after a row is created. I think that's just for annoying internal reasons like if you move a group its pointer changes so the row's parent needs to be updated. There might be a way around this but I think it's out of scope for this PR which has quite a lot of refactoring in it already!

DavidFair commented on qt/scientific_interfaces/ISISReflectometry/test/Reduction/MockGroup.h:
```diff
@@ -0,0 +1,52 @@
+#pragma once
+
+#include "../../../ISISReflectometry/Reduction/Group.h"
+#include <gmock/gmock.h>
```
You're missing a few headers in here:
`string`
`vector`
`boost/optional`
And from refl:
`Item.h`

Also `Group.h` should be `IGroup.h`

That way refactoring elsewhere in the codebase won't cause this to randomly break

DavidFair replied:
I'll mark this resolved, since it doesn't make sense to hold up the PR for a minor header change

DavidFair reviewed (changes requested):
Some minor points, overall a big improvement to the codebase :+1:
